File: test/test_reference.py
# Copyright 2026 Chi-kwan Chan
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import os
import subprocess
from pathlib import Path

import numpy as np
import pytest

pytestmark = pytest.mark.e2e


def reference_images():
    from cusmic.io import read_fits

    root = Path(__file__).parent / "data"
    return (
        read_fits(root / "input.fits.gz")[0],
        read_fits(root / "error.fits.gz")[0],
        read_fits(root / "reference.fits.gz")[0],
        read_fits(root / "reference.fits.gz", ext="CRMASK")[0],
    )


@pytest.mark.host
def test_saved_reference_lacosmic():
    from inspect import signature

    from cusmic import remove_cosmics as gpu_remove
    from lacosmic import remove_cosmics as cpu_remove

    image, error, expected, expected_mask = reference_images()
    assert signature(gpu_remove) == signature(cpu_remove)
    cleaned, mask = cpu_remove(image, 1, 5, 5, error=error, maxiter=4)
    np.testing.assert_array_equal(cleaned.view("uint64"), expected.view("uint64"))
    np.testing.assert_array_equal(mask, expected_mask)


def test_reference(cp, assert_pixels):
    from cusmic import remove_cosmics

    image, error, expected, expected_mask = reference_images()

    cleaned, mask = remove_cosmics(
        cp.asarray(image), error=cp.asarray(error),
        contrast=1, cr_threshold=5, neighbor_threshold=5, maxiter=4,
    )
    assert_pixels(cp.asnumpy(cleaned), expected)
    np.testing.assert_array_equal(cp.asnumpy(mask), expected_mask)


def test_cupy_cli_validation(cp, tmp_path):
    from click.testing import CliRunner
    from cusmic.__main__ import main
    from cusmic.io import read_fits

    source = Path(__file__).parent / "data/input.fits.gz"
    output = tmp_path / "copy.fits"
    args = [str(source), str(output), "--maxiter", "0"]
    result = CliRunner().invoke(main, args)
    assert result.exit_code == 0, result.output
    np.testing.assert_array_equal(read_fits(output)[0].view("uint64"),
                                  read_fits(source)[0].view("uint64"))
    assert not read_fits(output, ext="CRMASK")[0].any()
    assert CliRunner().invoke(main, args).exit_code != 0
    output = tmp_path / "invalid.fits"
    for options in (["--gain", "1", "--contrast", "nan"], ["--readnoise", "1"]):
        result = CliRunner().invoke(main, [str(source), str(output), *options])
        assert result.exit_code != 0 and not output.exists()


def test_scaled_fits_cli_parity(cp, tmp_path, assert_pixels):
    from astropy.io import fits
    from click.testing import CliRunner
    from cusmic.__main__ import main
    from cusmic.io import read_fits

    native = Path(__file__).resolve().parents[1] / os.environ.get(
        "CUSMIC_CUDA_CLI", "bin/cudasmic")
    if not native.exists():
        pytest.fail("Build bin/cudasmic for cross-implementation FITS checks", pytrace=False)

    data = np.ones((9, 9), dtype="int16")
    data[4, 4] = 10000
    hdu = fits.PrimaryHDU(data)
    hdu.header["BSCALE"] = 0.1
    hdu.header["BZERO"] = 10.2
    source, error = tmp_path / "scaled.fits", tmp_path / "error.fits"
    hdu.writeto(source)
    fits.PrimaryHDU(np.ones(data.shape)).writeto(error)

    for iterations in (0, 4):
        python_out = tmp_path / f"cupy-{iterations}.fits"
        cuda_out = tmp_path / f"cuda-{iterations}.fits"
        options = ["--error", str(error), "--maxiter", str(iterations)]
        result = CliRunner().invoke(main, [str(source), str(python_out), *options])
        assert result.exit_code == 0, result.output
        result = subprocess.run(
            [str(native), str(source), str(cuda_out), *options],
            capture_output=True, text=True,
        )
        assert result.returncode == 0, result.stderr

        python_pixels = read_fits(python_out)[0]
        cuda_pixels = read_fits(cuda_out)[0]
        assert_pixels(python_pixels, cuda_pixels, exact=iterations == 0)
        np.testing.assert_array_equal(
            read_fits(python_out, ext="CRMASK")[0],
            read_fits(cuda_out, ext="CRMASK")[0],
        )

File: test/test_reference.c
/*
 * Copyright 2026 Chi-kwan Chan
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "cusmic.h"
#include "io.h"
#include <float.h>
#include <inttypes.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int
main(int argc, char **argv)
{
	const char *dir = argc > 1 ? argv[1] : "test/data";
	const char *names[] = {"input.fits.gz", "error.fits.gz", "reference.fits.gz"};
	struct fits_image input = {0}, error = {0}, reference = {0}, flags = {0};
	struct fits_image *files[] = {&input, &error, &reference};
	struct cusmic_options o;
	struct cusmic_image im = {0};
	double *clean = NULL;
	uint8_t *mask = NULL;
	char path[1024], msg[256];
	int ret = 1, status;

	for (size_t i = 0; i < 3; ++i) {
		if (snprintf(path, sizeof(path), "%s/%s", dir, names[i]) >= (int)sizeof(path))
			goto out;
		status = read_fits(path, NULL, files[i]);
		if (status) {
			fits_report_error(stderr, status);
			goto out;
		}
	}
	if (snprintf(path, sizeof(path), "%s/reference.fits.gz", dir) >= (int)sizeof(path))
		goto out;
	status = read_fits(path, "CRMASK", &flags);
	if (status) {
		fits_report_error(stderr, status);
		goto out;
	}
	if (input.axes[0] != error.axes[0] || input.axes[1] != error.axes[1] ||
		input.axes[0] != reference.axes[0] || input.axes[1] != reference.axes[1] ||
		input.axes[0] != flags.axes[0] || input.axes[1] != flags.axes[1])
		goto out;

	size_t w = (size_t)input.axes[0], n = w * (size_t)input.axes[1];
	clean = malloc(n * sizeof(*clean));
	mask = malloc(n);
	if (!clean || !mask)
		goto out;
	im.data = input.data;
	im.error = error.data;
	im.width = w;
	im.height = (size_t)input.axes[1];
	cusmic_default_options(&o);
	o.contrast = 1;
	o.cr_threshold = o.neighbor_threshold = 5;
	if (remove_cosmics(&im, &o, clean, mask, msg, sizeof(msg))) {
		fprintf(stderr, "remove_cosmics: %s\n", msg);
		goto out;
	}

	for (size_t i = 0; i < n; ++i) {
		uint64_t got, want;
		memcpy(&got, clean + i, sizeof(got));
		memcpy(&want, reference.data + i, sizeof(want));
		double expected = reference.data[i];
		double tolerance = 32 * DBL_EPSILON * (1 + fabs(expected));
		if ((got != want && !(isfinite(clean[i]) && isfinite(expected) &&
			fabs(clean[i] - expected) <= tolerance)) ||
			mask[i] != (uint8_t)flags.data[i]) {
			fprintf(stderr, "reference differs at (%zu,%zu): pixels %.17g/%.17g"
				" (%016" PRIx64 "/%016" PRIx64 "), mask %u/%.0f\n",
				i % w, i / w, clean[i], expected, got, want, mask[i], flags.data[i]);
			goto out;
		}
	}
	puts("CUDA reference pixels are close and masks match exactly");
	ret = 0;
out:
	close_fits(&input);
	close_fits(&error);
	close_fits(&reference);
	close_fits(&flags);
	free(clean);
	free(mask);
	return ret;
}
